File: tutorials/007_plug_model_into_pipeline/plug_model_into_pipeline.py
#!/usr/bin/env python3
"""Two ways to plug a Model into a Session: direct vs. attached via ModelSessionOptions.

Usage:
  python3 plug_model_into_pipeline.py --mpk /path/to/yolo_v8s.tar.gz
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

try:
  import pyneat
except ImportError:
  sys.exit(
      "pyneat is not importable. Either Neat is not installed, or the venv is not activated.\n"
      "Run: source ~/pyneat/bin/activate"
  )


def main(argv: list[str]) -> int:
  ap = argparse.ArgumentParser(description=__doc__)
  ap.add_argument("--mpk", type=Path, required=True)
  args = ap.parse_args(argv[1:])

  model = pyneat.Model(str(args.mpk))

  # Pattern A: drop the model's own session group directly into a Session.
  # CORE LOGIC
  direct = pyneat.Session()
  direct.add(model.session())
  # END CORE LOGIC
  print(f"direct_session_size={model.session().size()}")

  # Pattern B: configure the session options for an attached upstream (e.g. a camera).
  sopt = pyneat.ModelSessionOptions()
  sopt.include_appsrc = False
  sopt.include_appsink = True
  sopt.upstream_name = "camera0"
  sopt.name_suffix = "_camera0"
  sopt.buffer_name = "camera0"

  # CORE LOGIC
  attached = pyneat.Session()
  attached.add(model.session(sopt))
  # END CORE LOGIC
  print(f"attached_session_built=True")
  return 0


if __name__ == "__main__":
  raise SystemExit(main(sys.argv))

File: tutorials/007_plug_model_into_pipeline/plug_model_into_pipeline.cpp
// Three Graph composition patterns: direct nodes, model.graph(), attached Graph.
//
// Usage:
//   tutorial_007_plug_model_into_pipeline [--model /path/to/model.tar.gz]

#include "neat.h"

#include <opencv2/core.hpp>

#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>

namespace fs = std::filesystem;

namespace {

bool get_arg(int argc, char** argv, const std::string& key, std::string& out) {
  for (int i = 1; i + 1 < argc; ++i) {
    if (key == argv[i]) {
      out = argv[i + 1];
      return true;
    }
  }
  return false;
}

} // namespace

int main(int argc, char** argv) {
  try {
    const int width = 224;
    const int height = 224;

    cv::Mat rgb(height, width, CV_8UC3, cv::Scalar(80, 40, 160));
    if (!rgb.isContinuous())
      rgb = rgb.clone();

    // Pattern 1: build a Graph by adding Input/Output nodes directly.
    simaai::neat::InputOptions in;
    in.format = "RGB";
    in.width = width;
    in.height = height;
    in.depth = 3;
    in.do_timestamp = true;

    // CORE LOGIC
    simaai::neat::Graph direct;
    direct.add(simaai::neat::nodes::Input(in));
    direct.add(simaai::neat::nodes::Output());
    // END CORE LOGIC

    std::string model_path;
    if (get_arg(argc, argv, "--model", model_path) && fs::exists(model_path)) {
      simaai::neat::Model model(model_path);

      // Pattern 2: ask model.graph() to include explicit public Input/Output boundaries.
      // STEP model-graph
      simaai::neat::Model::RouteOptions runnable_opt;
      runnable_opt.include_input = true;
      runnable_opt.include_output = true;
      // CORE LOGIC
      simaai::neat::Graph from_model;
      from_model.add(model.graph(runnable_opt));
      // END CORE LOGIC
      // END STEP
      std::cout << "model_graph_backend=\n" << from_model.describe_backend() << "\n";

      // Pattern 3: attach the model under an upstream name with custom options.
      // STEP route-options
      simaai::neat::Model::RouteOptions sopt;
      sopt.include_input = false;
      sopt.include_output = true;
      sopt.upstream_name = "camera0";
      sopt.name_suffix = "_camera0";
      sopt.buffer_name = "camera0";
      // END STEP

      // CORE LOGIC
      // STEP attached-graph
      simaai::neat::Graph attached;
      attached.add(model.graph(sopt));
      // END STEP
      // END CORE LOGIC
      std::cout << "attached_graph_backend=\n" << attached.describe_backend() << "\n";
    }

    auto run = direct.build(std::vector<cv::Mat>{rgb}, simaai::neat::RunMode::Sync);
    simaai::neat::TensorList out = run.run(std::vector<cv::Mat>{rgb}, /*timeout_ms=*/1000);

    if (out.empty())
      throw std::runtime_error("direct Graph output missing tensor");
    std::cout << "direct_rank=" << out.front().shape.size() << "\n";
    std::cout << "[OK] 007_plug_model_into_pipeline\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "[FAIL] " << e.what() << "\n";
    return 1;
  }
}
